contracts/farm/LiquidityMining.sol

```diff
@@ -0,0 +1,567 @@
+	 * @param _userAddress The address of the user, to send SOV in its behalf.
+	 */
+	function _transferReward(
+		UserInfo storage _user,
+		address _userAddress,
+		bool _isClaimingReward
+	) internal {
+		uint256 userAccumulatedReward = _user.accumulatedReward;
+
+		/// @dev Transfer if enough SOV balance on this LM contract.
+		uint256 balance = SOV.balanceOf(address(this));
+		if (balance >= userAccumulatedReward) {
+			totalUsersBalance = totalUsersBalance.sub(userAccumulatedReward);
+			_user.accumulatedReward = 0;
+
+			/// @dev Instead of transferring the reward to the LP (user),
+			///   deposit it into lockedSOV vault contract, but first
+			///   SOV deposit must be approved to move the SOV tokens
+			///   from this LM contract into the lockedSOV vault.
+			SOV.approve(address(lockedSOV), userAccumulatedReward);
+			lockedSOV.deposit(_userAddress, userAccumulatedReward, basisPoint);
+
+			if (_isClaimingReward) {
+				lockedSOV.withdrawAndStakeTokensFrom(_userAddress);
+			}
+
+			/// @dev Event log.
+			emit RewardClaimed(_userAddress, userAccumulatedReward);
+		}
+	}
+
+	/**
+	 * @notice withdraws pool tokens without transferring reward tokens
+	 * @param _poolToken the address of pool token
+	 * @dev EMERGENCY ONLY
+	 */
+	function emergencyWithdraw(address _poolToken) public {
+		uint256 poolId = _getPoolId(_poolToken);
+		PoolInfo storage pool = poolInfoList[poolId];
+		UserInfo storage user = userInfoMap[poolId][msg.sender];
+
+		uint256 userAmount = user.amount;
+		user.amount = 0;
+		user.rewardDebt = 0;
+		user.accumulatedReward = 0;
+		pool.poolToken.safeTransfer(address(msg.sender), userAmount);
+		emit EmergencyWithdraw(msg.sender, _poolToken, userAmount);
+	}
+
+	/**
+	 * @notice returns pool id
+	 * @param _poolToken the address of pool token
+	 */
+	function getPoolId(address _poolToken) public view returns (uint256) {
+		return _getPoolId(_poolToken);
+	}
+
+	function _getPoolId(address _poolToken) internal view returns (uint256) {
+		uint256 poolId = poolIdList[_poolToken];
+		require(poolId > 0, "Pool token not found");
+		return poolId - 1;
+	}
+
+	/**
+	 * @notice returns count of pool tokens
+	 */
+	function getPoolLength() external view returns (uint256) {
+		return poolInfoList.length;
+	}
+
+	/**
+	 * @notice returns list of pool token's info
+	 */
+	function getPoolInfoList() external view returns (PoolInfo[] memory) {
+		return poolInfoList;
+	}
+
+	/**
+	 * @notice returns pool info for the given token
+	 * @param _poolToken the address of pool token
+	 */
+	function getPoolInfo(address _poolToken) external view returns (PoolInfo memory) {
+		uint256 poolId = _getPoolId(_poolToken);
+		return poolInfoList[poolId];
+	}
+
+	/**
+	 * @notice returns list of [amount, accumulatedReward] for the given user for each pool token
+	 * @param _user the address of the user
+	 */
+	function getUserBalanceList(address _user) external view returns (uint256[2][] memory) {
```

Comment:
Is it not necessary to update the reward balance first?

Reply:
We shouldn't update a storage in getters.
All necessary calculations are made in this function.